### ISSUE TITLE
[TypeError: Cannot read property 'collections' of undefined] when calling model.collection.add

### ISSUE THREAD

**Issue by joeandrews** [balderdashy/waterline#1040]
With sails v0.11.0 and latest sails-mongo I am getting this error. I think the error is coming from waterline any ideas on where to start ?

**Comment by tjwebb** [comment]
You can start by sharing your code

**Comment by joeandrews** [comment]
Good idea. 

The user.save() call is throwing the error, I have also included my model definitions.

My understanding is that the collection.add() helper method queues up functions to run on save to actually create the required associations.

```

user.outstandingCredits.add(referralCode.id);

return user.save();

```


===============================

User Model

```
module.exports = {
	attributes: {
		validAddress: {
			type: 'boolean'
		},
		creditCard: {
			type: 'boolean'
		},
		referralCode: {
			model: 'referral'
		},
		redeemedCodes: {
			collection: 'referral',
			via: 'redemptions'
		},
		isAdmin: {
			type: 'boolean'
		},
		name: {
			type: 'string',
			required: true
		},
		phone: {
			type: 'string',
			required: true,
			unique: true
		},
		outstandingCredits: {
			collection: 'referral'
		},
		confirmationCode: {
			type: 'integer',
			required: true
		},
		password: {
			type: 'string',
			required: true
		},
		addresses: {
			type: 'json',
		},
		email: {
			type: 'email',
			required: true,
			unique: true
		}
}
```

Referral Model 

```
attributes: {
		
		customerText: {
			type: 'string',
			defaultsTo: 'Referral Discount'
		},
		code: {
			type: 'string',
			required: true,
			unique: true
		},
		expires: {
			type: 'date'
		},
		onlyNewUsers: {
			type: 'boolean',
			defaultsTo: true
		},
		user: {
			model: 'user'
		},
		referrerCredit: {
			type: 'string',
			defaultsTo: '10'
		},
		dollarValue: {
			type: 'string',
			defaultsTo: '10'
		},
		redemptions: {
			collection: 'user',
			via: 'redeemedCodes'
		}
}
```

The error started appearing after I ran a fresh NPM install with latest Sails and latest Sails-Mongo. 

Thanks 

Joe

**Comment by dmarcelino** [comment]
Hi @joeandrews,

Your `user.outstandingCredits` model definition is missing the `via` keyword (even though I don't think that's causing the issue).

BTW, which version of `waterline` are you using? And can you provide the stack trace?

**Comment by joeandrews** [comment]
Hi @dmarcelino,

Thanks for the reply. Just installed waterline@0.10.23, was previously using the one that is bundled with sails.  I also just added the following to the user definition.

```
via:'id'
```

The error is still the same.

```
>> Trace: [TypeError: Cannot read property 'collections' of undefined]
>>     at /Users/joeandrews/Documents/Code/Radish/build/backend/api/services/ReferralService.js:211:17
>>     at tryCatcher (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/util.js:24:31)
>>     at Promise._settlePromiseFromHandler (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:454:31)
>>     at Promise._settlePromiseAt (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:530:18)
>>     at Promise._settlePromises (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:646:14)
>>     at Async._drainQueue (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/async.js:177:16)
>>     at Async._drainQueues (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/async.js:187:10)
>>     at Async.drainQueues (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/async.js:15:14)
>>     at process._tickDomainCallback (node.js:463:13)

```

**Comment by dmarcelino** [comment]
Shouldn't it be `via: 'user'` to match attribute:
```js
        user: {
            model: 'user'
        },
```
hence forming a [one-to-many association](https://github.com/balderdashy/waterline-docs/blob/master/associations.md#one-to-many-associations)?

The stack trace doesn't really go deep enough... :(

**Comment by joeandrews** [comment]
@dmarcelino 

The intention of the schema is to be able to assign a user many outstanding credits, but I do not need to keep track of the outstanding credits on the referral code. A referral code can be given to many users through outstandingCredits, but is owned by one via the user attribute on the code.

How can I get more info from the stack trace?

**Comment by dmarcelino** [comment]
The thing is, waterline does not explicitly support *one way many* associations and, according to [stackoverflow #30198701](http://stackoverflow.com/questions/30198701/sails-js-is-there-intended-support-for-a-one-way-many-association), creating an association on one model and not a corresponding association in the other will result in a many-to-many association being created which is not ideal.

Not sure to be honest.

**Comment by joeandrews** [comment]
@dmarcelino Thanks for the info.

Even if I create the association, as a many to many association with the following in both schema's I run into the same issue.

```

outstandingCredits:{
    	collection: 'referral',
	via: 'outstanding',
        dominant:true

}

```

```

outstanding:{
    	collection: 'user',
	via: 'outstandingCredits'

}

```
I will keep digging into this.

**Comment by dmarcelino** [comment]
I feared that wasn't the root issue...

Regarding the stack trace, can you try logging the exception earlier? For example:
```js
return user.save(function(err){
  if(err) console.error('save failed:', err);
});
```

**Comment by joeandrews** [comment]
Progress!
```
>> Unhandled rejection TypeError: Cannot read property 'collections' of undefined
>>     at /Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/utils/nestedOperations/reduceAssociations.js:50:33
>>     at Array.forEach (native)
>>     at Object.module.exports [as reduceAssociations] (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/utils/nestedOperations/reduceAssociations.js:25:23)
>>     at /Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/associationMethods/update.js:65:36
>>     at Array.forEach (native)
>>     at new module.exports (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/associationMethods/update.js:45:8)
>>     at Array.async.auto.updateRecord [as 1] (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:73:41)
>>     at /Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:640:38
>>     at _each (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:57:9)
>>     at Object.async.auto (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/async/lib/async.js:600:9)
>>     at new module.exports (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:42:9)
>>     at prototypeFns.save (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/lib/waterline/model/index.js:34:14)
>>     at /Users/joeandrews/Documents/Code/Radish/build/backend/api/services/ReferralService.js:202:14
>>     at tryCatcher (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/util.js:24:31)
>>     at Promise._settlePromiseFromHandler (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:454:31)
>>     at Promise._settlePromiseAt (/Users/joeandrews/Documents/Code/Radish/build/backend/node_modules/sails/node_modules/waterline/node_modules/bluebird/js/main/promise.js:530:18)

```

**Comment by dmarcelino** [comment]
Much better!

**Comment by joeandrews** [comment]
This commit is the issue 

https://github.com/balderdashy/waterline/commit/3a8bab84c0e677c1b08b77a61f3ae08bde2e46de?diff=split

**Comment by dmarcelino** [comment]
Good find, hopefully @particlebanana will comment on it.

I still can't figure out why this change breaks your case and manages to pass the tests in https://github.com/balderdashy/waterline-adapter-tests/blob/master/interfaces/associations/manyToMany/add.js. I can't see any obvious difference between what you are doing and the tests...

**Comment by gblock0** [comment]
Yes

**Comment by dmarcelino** [comment]
@gblock0, are you also experiencing this? Can you share your code and models? It may helps us understand how your case scenarios differ from the test scenarios.

**Comment by gblock0** [comment]
@dmarcelino I am working with @joeandrews. Sorry I should have specified. We tried downgrading to 0.10.22 and that fixed the issue we were seeing.

**Comment by dmarcelino** [comment]
LOL! That makes sense, cool :)

**Comment by dmarcelino** [comment]
@joeandrews, @gblock0, if you guys can come up with a new test in [waterline-adapter-tests manyToMany/add.js](https://github.com/balderdashy/waterline-adapter-tests/blob/master/interfaces/associations/manyToMany/add.js) that reproduces your issue, it would help a lot!

**Comment by joeandrews** [comment]
will do this weekend!

**Comment by devinivy** [comment]
I just ran into this in a sails project!  No more info to report, but it's definitely an existing bug.  Might have to do with lifecycle.

**Comment by devinivy** [comment]
Able to fix by changing [this line](https://github.com/balderdashy/waterline/blob/317cdf0173990face8afe67b336bc6e6d0fe5630/lib/waterline/model/lib/associationMethods/update.js#L65)
```node
    var reduced = nestedOperations.reduceAssociations(collection.identity, collection.waterline.schema, vals);

```
to
```node
    var reduced = nestedOperations.reduceAssociations.call(collection, collection.identity, collection.waterline.schema, vals);
```

Don't have time to write a PR, but anyone feel free!  We should be able to figure out roughly what went wrong (what may need testing) by where that code change occurs.

**Comment by dmarcelino** [comment]
Potential fix in PR #1057, it could use a failing test, though.
